src/ccsr/tls.rs

```diff
@@ -9,49 +9,53 @@
 
 //! TLS certificates and identities.
 
+use openssl::pkcs12::Pkcs12;
+use openssl::pkey::PKey;
+use openssl::stack::Stack;
+use openssl::x509::X509;
 use serde::{Deserialize, Serialize};
 
-// Encodes the type of certificate file, as well as the certificate's bytes. In
-// the case of der certificates, it also stores the password.
-#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
-pub(crate) enum CertDetails {
-    Pem(Vec<u8>),
-    Der(Vec<u8>, String),
-}
-
 /// A [Serde][serde]-enabled wrapper around [`reqwest::Identity`].
 #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
 pub struct Identity {
-    pub(crate) cert: CertDetails,
+    der: Vec<u8>,
+    pass: String,
 }
 
 impl Identity {
-    /// Wraps [`reqwest::Identity::from_pem`].
-    pub fn from_pem(pem: &[u8]) -> Result<Self, reqwest::Error> {
-        let _ = reqwest::Identity::from_pem(&pem)?;
-        Ok(Identity {
-            cert: CertDetails::Pem(pem.into()),
-        })
+    /// Reimplements [`reqwest::Identity::from_pem`] in terms of OpenSSL.
+    ///
+    /// The implementation in reqwest requires rustls.
+    pub fn from_pem(pem: &[u8]) -> Result<Self, openssl::error::ErrorStack> {
```

Comment:
Probably immaterial or unadvisable, but I'd expect the arguments of `from_pem` and `from_pkcs12_der` to have the same argument types, i.e. either both or neither of them would be references. For my own edification, is there a reason you've gone with these signatures?

Reply:
Yeah, it's definitely a bit weird, isn't it? But there is a general consideration in Rust, moreso than in other languages, to give the caller explicit control over clones. https://rust-lang.github.io/api-guidelines/flexibility.html#c-caller-control

In particular, `Certificate` ultimately needs an owned DER-encoded string and an owned password. If `from_pkcs12_der` were to take its inputs by reference, it would have to unconditionally clone them. By taking them by value, we give the caller a chance to provide us with an owned buffer that they no longer need. (And, it turns out, in this specific case, we're willing to give up ownership in each call site in this case.) If the caller needs to retain ownership, they can call `clone`, and we're no worse off.

Exactly the opposite applies to `from_pem`. Since we're going to be converting the PEM bytes into DER bytes, if you gave us ownership of the PEM bytes, we'd have to drop that buffer on the floor at the end of the function, once we'd constructed the DER bytes. But by taking the argument by reference, we let the caller potentially do something else with the PEM bytes after `from_pem` returns.

This code isn't performance sensitive, so it doesn't really matter on that front, but it's a habit I've been trying to build nonetheless. Many Rust APIs will implicitly tell you how much cloning will occur just based on the type signatures involved.